group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -10921,6 +10823,1158 @@ public void testLastConsumerProtocolMemberRebalanceTimeoutInConsumerGroup() {
+                    }
+                }),
+                RecordHelpers.newGroupEpochRecord(groupId, 11),
+
+                RecordHelpers.newTargetAssignmentRecord(groupId, memberId, assignor.targetPartitions(memberId)),
+                RecordHelpers.newTargetAssignmentRecord(groupId, newMemberId, assignor.targetPartitions(newMemberId)),
+
+                RecordHelpers.newTargetAssignmentEpochRecord(groupId, 11),
+                RecordHelpers.newCurrentAssignmentRecord(groupId, expectedMember)
+            );
+            assertRecordsEquals(expectedRecords.subList(0, 3), secondJoinResult.records.subList(0, 3));
+            assertUnorderedListEquals(expectedRecords.subList(3, 5), secondJoinResult.records.subList(3, 5));
+            assertRecordsEquals(expectedRecords.subList(5, 7), secondJoinResult.records.subList(5, 7));
+
+            if (replaySuccessfully) {
+                secondJoinResult.appendFuture.complete(null);
+                assertTrue(secondJoinResult.joinFuture.isDone());
+                assertEquals(
+                    new JoinGroupResponseData()
+                        .setMemberId(newMemberId)
+                        .setGenerationId(11)
+                        .setProtocolType(ConsumerProtocol.PROTOCOL_TYPE)
+                        .setProtocolName("range"),
+                    secondJoinResult.joinFuture.get()
+                );
+
+                context.assertSessionTimeout(groupId, newMemberId, request.sessionTimeoutMs());
+                context.assertSyncTimeout(groupId, newMemberId, request.rebalanceTimeoutMs());
+            } else {
+                secondJoinResult.appendFuture.completeExceptionally(new NotLeaderOrFollowerException());
+                assertEquals(
+                    new JoinGroupResponseData()
+                        .setErrorCode(Errors.NOT_COORDINATOR.code()),
+                    secondJoinResult.joinFuture.get()
+                );
+                context.assertNoSessionTimeout(groupId, newMemberId);
+                context.assertNoSyncTimeout(groupId, newMemberId);
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testJoiningConsumerGroupWithNewStaticMember(boolean replaySuccessfully) throws Exception {
+        String groupId = "group-id";
+        Uuid fooTopicId = Uuid.randomUuid();
+        String fooTopicName = "foo";
+        Uuid barTopicId = Uuid.randomUuid();
+        String barTopicName = "bar";
+
+        for (short version = ConsumerProtocolSubscription.LOWEST_SUPPORTED_VERSION; version <= ConsumerProtocolSubscription.HIGHEST_SUPPORTED_VERSION; version++) {
+            String memberId = Uuid.randomUuid().toString();
+            String instanceId = "instance-id-" + version;
+            MockPartitionAssignor assignor = new MockPartitionAssignor("range");
+            assignor.prepareGroupAssignment(new GroupAssignment(Collections.emptyMap()));
```

Comment:
This is a little weird, no? It may be better to have a target assignment for the group.

Reply:
This is because the new static member uses the newly generated member id. We don't know what the member id will be when preparing the assignment. 

We might need another test that makes the existing static member join and compute a new target assignment.